### ISSUE TITLE
High-memory consumption in AMG

### ISSUE THREAD

**Issue by constantinpape** [computational-cell-analytics/micro-sam#188]
The AMG needs a lot of memory, which can lead to severe drops in performance, especially when using MPS. See #183 for details.
The underlying reason is that we store all the masks that are computed from the point grid in `initialize`, and then only filter out low quality masks (measured by the predicted IOU score and the stability score) in `generate`.

While the masks are stored in RLE after `initialize` (and so don't need much storage), they are represented as binary masks in the intermediate computation steps. So by default there will be 32 x 32 x 3 masks. (32 is the default number of points per dimension, and for each points 3 masks will be predicted). Hence they take up some memory, and the computations run on top (computing the stability scores, RLEs etc.) will take up some more.

There are two possible ways to decrease this memory requirement:
- See which of the intermediate computation takes up most memory and look into more memory efficient implementations (for this we would need to run memory profiling; I am not so familiar with the best practices for this in python)
- Discard masks of low quality early. I.e. already include thresholds for the `iou_score` and `stability_score` (set to a fairly conservative value that keeps all relevant masks) in `initialize` and use it to pre-filter.

cc @GenevieveBuckley

**Comment by GenevieveBuckley** [comment]
> Do you know any tool for memory line profiling?

I have done a little bit of work with memory profiling before, but not lots. 

* [memray](https://github.com/bloomberg/memray) - this is a new project I haven't used before, and it looks really interesting. 
    * I also think it's cool that there is a [pytest-memray](https://github.com/bloomberg/pytest-memray) plugin. I'm not sure if it's a good fit for micro-sam, but I haven't seem memory profilers hooked into pytest in this way before, and it seems like a useful thing to have in some circumstances.
* [Fil](https://github.com/pythonspeed/filprofiler) - I have used this before, but don't remember a lot about it. [Here is an example](https://blog.dask.org/2021/03/11/dask_memory_usage#fil), and [the Fil blogpost](https://pythonspeed.com/articles/memory-profiler-data-scientists/).
* [memory-profiler](https://github.com/pythonprofilers/memory_profiler) - I've used this before, although a note was added last year to the README to say it is no longer actively maintained. So it might be better to use a newer option like [memray](https://github.com/bloomberg/memray) instead if we can.
* [guppy3](https://github.com/zhuyifei1999/guppy3) - might also be an option, I've never used it before.

**Comment by GenevieveBuckley** [comment]
I'll take a quick look at it today or tomrrow

**Comment by constantinpape** [comment]
I looked into this a bit more and tried 3 things:
1. implementing the RLE computation in numba in #190. This is slower than the current implementation, but may save memory (I haven't profiled memory yet).
2. filtering unlikely masks in `initialize` in #191. This does not affect runtime signifcantly, but may save memory (again not yet mem profiled).
3. implementing the RLE computation in C++. This does not negatively affect runtime and may also save memory (I would assume that the memory saving is the same as the numba implementation).

@GenevieveBuckley could you check if options 1 or 2 actually save memory? Depending on this we can see how to continue. Note that 3 cannot be tested yet because the C++ code is unreleased, but I think we can just extrapolate the results from 3 to it (and I have put the usage of it in comments in #190).

Thanks!

**Comment by GenevieveBuckley** [comment]
I ran [memray](https://bloomberg.github.io/memray/) on a single iteration of the amg benchmark from the `benchmark.py` file. 

I emailed Constantin a copy of the output .bin files yesterday. Looking at the flamegraphs, I think the biggest memory consumer is loading the model itself, and the next biggest is the `mask_to_rle` computations.

Here are the memory profiler summary stats from the current dev branch: 
<details>
<summary>vit_b (click to expand)</summary>

```
memray stats memray-output-detailed-amg-vit_b.bin
📏 Total allocations:
	54383659

📦 Total memory allocated:
	8.280GB

📊 Histogram of allocation size:
	min: 0.000B
	-----------------------------------------------
	< 6.000B   :    75669 ▇
	< 37.000B  : 17694397 ▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 225.000B : 35440641 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 1.344KB  :   933491 ▇
	< 8.187KB  :   180202 ▇
	< 49.868KB :    49047 ▇
	< 303.748KB:     6924 ▇
	< 1.807MB  :     3151 ▇
	< 11.005MB :      116 ▇
	<=67.031MB :       21 ▇
	-----------------------------------------------
	max: 67.031MB

📂 Allocator type distribution:
	 PYMALLOC_MALLOC: 49356617
	 MALLOC: 2661871
	 PYMALLOC_REALLOC: 928414
	 CALLOC: 634006
	 POSIX_MEMALIGN: 464982
	 PYMALLOC_CALLOC: 321287
	 REALLOC: 16472
	 MMAP: 10

🥇 Top 5 largest allocating locations (by size):
	- mask_to_rle_pytorch:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:91 -> 1.048GB
	- mask_to_rle_pytorch:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:84 -> 1.047GB
	- _deepcopy_list:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/copy.py:206 -> 878.503MB
	- deepcopy:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/copy.py:138 -> 877.428MB
	- deepcopy:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/copy.py:144 -> 877.428MB

🥇 Top 5 largest allocating locations (by number of allocations):
	- deepcopy:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/copy.py:138 -> 12778469
	- deepcopy:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/copy.py:144 -> 12778467
	- deepcopy:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/copy.py:137 -> 12778467
	- _compile_bytecode:<frozen importlib._bootstrap_external>:672 -> 1352323
	- mask_to_rle_pytorch:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:96 -> 971485
```

</details>

<details>
<summary>vit_l (click to expand)</summary>

```
memray stats memray-output-amg-vit_l.bin
📏 Total allocations:
	4365278

📦 Total memory allocated:
	7.470GB

📊 Histogram of allocation size:
	min: 0.000B
	----------------------------------------------
	< 6.000B   :   76036 ▇
	< 37.000B  : 1903764 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 225.000B : 1414510 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 1.344KB  :  720471 ▇▇▇▇▇▇▇▇▇▇
	< 8.187KB  :  184257 ▇▇▇
	< 49.868KB :   55143 ▇
	< 303.748KB:    7500 ▇
	< 1.807MB  :    3364 ▇
	< 11.005MB :      68 ▇
	<=67.031MB :     165 ▇
	----------------------------------------------
	max: 67.031MB

📂 Allocator type distribution:
	 MALLOC: 2989224
	 CALLOC: 736168
	 POSIX_MEMALIGN: 466225
	 REALLOC: 173474
	 MMAP: 187

🥇 Top 5 largest allocating locations (by size):
	- load_tensor:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/serialization.py:1357 -> 1.164GB
	- __init__:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/modules/linear.py:96 -> 1.140GB
	- mask_to_rle_pytorch:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:84 -> 1.047GB
	- mask_to_rle_pytorch:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:91 -> 1.047GB
	- _deepcopy_list:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/copy.py:206 -> 1.013GB

🥇 Top 5 largest allocating locations (by number of allocations):
	- <stack trace unavailable> -> 358211
	- forward:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/modules/linear.py:114 -> 311930
	- cat:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/segment_anything/utils/amg.py:64 -> 254343
	- _call_with_frames_removed:<frozen importlib._bootstrap>:241 -> 206986
	- layer_norm:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/functional.py:2543 -> 160940
```

</details>

<details>
<summary>vit_h (click to expand)</summary>
```
memray stats memray-output-amg-vit_h.bin
📏 Total allocations:
	4439043

📦 Total memory allocated:
	10.015GB

📊 Histogram of allocation size:
	min: 0.000B
	----------------------------------------------
	< 6.000B   :   78229 ▇▇
	< 37.000B  : 1938693 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 225.000B : 1445049 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 1.344KB  :  733839 ▇▇▇▇▇▇▇▇▇▇
	< 8.187KB  :  180472 ▇▇▇
	< 49.868KB :   48912 ▇
	< 303.748KB:   10168 ▇
	< 1.807MB  :    3384 ▇
	< 11.005MB :      84 ▇
	<=67.031MB :     213 ▇
	----------------------------------------------
	max: 67.031MB

📂 Allocator type distribution:
	 MALLOC: 3024885
	 CALLOC: 781845
	 POSIX_MEMALIGN: 466965
	 REALLOC: 165160
	 MMAP: 188

🥇 Top 5 largest allocating locations (by size):
	- load_tensor:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/serialization.py:1357 -> 2.388GB
	- __init__:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/modules/linear.py:96 -> 2.359GB
	- _deepcopy_list:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/copy.py:206 -> 1.065GB
	- mask_to_rle_pytorch:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:84 -> 1.047GB
	- mask_to_rle_pytorch:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:91 -> 1.047GB

🥇 Top 5 largest allocating locations (by number of allocations):
	- <stack trace unavailable> -> 365821
	- forward:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/modules/linear.py:114 -> 320654
	- cat:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/segment_anything/utils/amg.py:64 -> 254397
	- _call_with_frames_removed:<frozen importlib._bootstrap>:241 -> 208003
	- layer_norm:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/functional.py:2543 -> 164208
```

</details>

**Comment by constantinpape** [comment]
Thanks for posting it here too @GenevieveBuckley. If I have some time later I will see if I get to run the mem profiling to run myself to also compare this for #190 and #191. (But it would still be great if you can run the profilers for those as well at some point :))

**Comment by GenevieveBuckley** [comment]
I left [this comment](https://github.com/computational-cell-analytics/micro-sam/pull/190#issuecomment-1722641524) on the rle-numba PR, but since it refers to all the current branches involving memory use, I'm copying the contents here too.

> > Unfortunately the current attempt is slower than the previous implementation:
> 
> That's interesting, I don't see the same results.
> 
> On my machine, the rle-numba branch is slightly faster than the current dev branch. But the prefilter amg branch with the vit_h model did exceed the available 16GB memory, which is why I think we see a slowdown with the mps backend there. I imagine if more memory was available it wouldn't be slower.
> 
> 3 iterations of the amg benchmark function, average runtime reported:
> | model   | device   | branch         | benchmark   |   runtime |
> |:--------|:---------|:---------------|:------------|----------:|
> | vit_h   | cpu      | dev            | amg         |   37.7422 |
> | vit_h   | mps      | dev            | amg         |   33.8933 |
> | vit_h   | cpu      | rle-numba      | amg         |   35.9172 |
> | vit_h   | mps      | rle-numba      | amg         |   33.4212 |
> | vit_h   | cpu      | prefilter-amg  | amg         |   33.2118 |
> | vit_h   | mps      | prefilter-amg  | amg         |   43.8554 | 
> | vit_h   | cpu      | numba+prefilter| amg         |   34.6391 |
> | vit_h   | mps      | numba+prefilter| amg         |   42.9275 |
> 
> *The `numba+prefilter` branch here is me merging PR [Filter unlikely masks in AMG #191](https://github.com/computational-cell-analytics/micro-sam/pull/191)  and [Implement RLE computation with numba #190](https://github.com/computational-cell-analytics/micro-sam/pull/190) just to see what the combination would do.

**Comment by GenevieveBuckley** [comment]
Here are the memray summary statistics for the amg memory benchmarks.


<details><summary>Details: dev branch</summary>
<p>

<details><summary>Details: memray stats memray-output-amg-vit_b.bin</summary>
<p>

```
📏 Total allocations:
	4218946

📦 Total memory allocated:
	5.585GB

📊 Histogram of allocation size:
	min: 0.000B
	----------------------------------------------
	< 6.000B   :   72175 ▇
	< 37.000B  : 1846375 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 225.000B : 1362981 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 1.344KB  :  697829 ▇▇▇▇▇▇▇▇▇▇
	< 8.187KB  :  180139 ▇▇▇
	< 49.868KB :   49049 ▇
	< 303.748KB:    6924 ▇
	< 1.807MB  :    3337 ▇
	< 11.005MB :     116 ▇
	<=67.031MB :      21 ▇
	----------------------------------------------
	max: 67.031MB

📂 Allocator type distribution:
	 MALLOC: 2929856
	 CALLOC: 659165
	 POSIX_MEMALIGN: 465115
	 REALLOC: 164614
	 MMAP: 196

🥇 Top 5 largest allocating locations (by size):
	- mask_to_rle_pytorch:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:84 -> 1.047GB
	- mask_to_rle_pytorch:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:91 -> 1.047GB
	- _deepcopy_list:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/copy.py:206 -> 867.996MB
	- load_tensor:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/serialization.py:1357 -> 357.687MB
	- __init__:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/modules/linear.py:96 -> 339.159MB

🥇 Top 5 largest allocating locations (by number of allocations):
	- <stack trace unavailable> -> 341937
	- forward:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/modules/linear.py:114 -> 302202
	- cat:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/segment_anything/utils/amg.py:64 -> 248369
	- _call_with_frames_removed:<frozen importlib._bootstrap>:241 -> 206988
	- layer_norm:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/functional.py:2543 -> 152427
```

</p>
</details> 

<details><summary>Details: memray stats memray-output-amg-vit_l.bin</summary>
<p>

```
📏 Total allocations:
	4365278

📦 Total memory allocated:
	7.470GB

📊 Histogram of allocation size:
	min: 0.000B
	----------------------------------------------
	< 6.000B   :   76036 ▇
	< 37.000B  : 1903764 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 225.000B : 1414510 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 1.344KB  :  720471 ▇▇▇▇▇▇▇▇▇▇
	< 8.187KB  :  184257 ▇▇▇
	< 49.868KB :   55143 ▇
	< 303.748KB:    7500 ▇
	< 1.807MB  :    3364 ▇
	< 11.005MB :      68 ▇
	<=67.031MB :     165 ▇
	----------------------------------------------
	max: 67.031MB

📂 Allocator type distribution:
	 MALLOC: 2989224
	 CALLOC: 736168
	 POSIX_MEMALIGN: 466225
	 REALLOC: 173474
	 MMAP: 187

🥇 Top 5 largest allocating locations (by size):
	- load_tensor:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/serialization.py:1357 -> 1.164GB
	- __init__:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/modules/linear.py:96 -> 1.140GB
	- mask_to_rle_pytorch:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:84 -> 1.047GB
	- mask_to_rle_pytorch:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:91 -> 1.047GB
	- _deepcopy_list:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/copy.py:206 -> 1.013GB

🥇 Top 5 largest allocating locations (by number of allocations):
	- <stack trace unavailable> -> 358211
	- forward:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/modules/linear.py:114 -> 311930
	- cat:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/segment_anything/utils/amg.py:64 -> 254343
	- _call_with_frames_removed:<frozen importlib._bootstrap>:241 -> 206986
	- layer_norm:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/functional.py:2543 -> 160940
```

</p>
</details> 

<details><summary>Details: memray stats memray-output-amg-vit_h.bin</summary>
<p>


```
📏 Total allocations:
	4439043

📦 Total memory allocated:
	10.015GB

📊 Histogram of allocation size:
	min: 0.000B
	----------------------------------------------
	< 6.000B   :   78229 ▇▇
	< 37.000B  : 1938693 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 225.000B : 1445049 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 1.344KB  :  733839 ▇▇▇▇▇▇▇▇▇▇
	< 8.187KB  :  180472 ▇▇▇
	< 49.868KB :   48912 ▇
	< 303.748KB:   10168 ▇
	< 1.807MB  :    3384 ▇
	< 11.005MB :      84 ▇
	<=67.031MB :     213 ▇
	----------------------------------------------
	max: 67.031MB

📂 Allocator type distribution:
	 MALLOC: 3024885
	 CALLOC: 781845
	 POSIX_MEMALIGN: 466965
	 REALLOC: 165160
	 MMAP: 188

🥇 Top 5 largest allocating locations (by size):
	- load_tensor:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/serialization.py:1357 -> 2.388GB
	- __init__:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/modules/linear.py:96 -> 2.359GB
	- _deepcopy_list:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/copy.py:206 -> 1.065GB
	- mask_to_rle_pytorch:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:84 -> 1.047GB
	- mask_to_rle_pytorch:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:91 -> 1.047GB

🥇 Top 5 largest allocating locations (by number of allocations):
	- <stack trace unavailable> -> 365821
	- forward:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/modules/linear.py:114 -> 320654
	- cat:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/segment_anything/utils/amg.py:64 -> 254397
	- _call_with_frames_removed:<frozen importlib._bootstrap>:241 -> 208003
	- layer_norm:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/functional.py:2543 -> 164208
(test-micro-sam-mps-nightly) genevieb@192-168-1-102 development %

```

</p>
</details> 


</p>
</details> 


<details><summary>Details: prefilter-amg branch</summary>
<p>

<details><summary>Details: memray stats memray-output-prefilter-amg-vit_b.bin </summary>
<p>

```
📏 Total allocations:
	6686407

📦 Total memory allocated:
	4.289GB

📊 Histogram of allocation size:
	min: 0.000B
	----------------------------------------------
	< 5.000B   :   73906 ▇
	< 34.000B  : 3118725 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 203.000B : 2154728 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 1.167KB  : 1085524 ▇▇▇▇▇▇▇▇▇
	< 6.864KB  :  211312 ▇▇
	< 40.368KB :   37664 ▇
	< 237.376KB:    2179 ▇
	< 1.363MB  :    2231 ▇
	< 8.015MB  :      69 ▇
	<=47.131MB :      69 ▇
	----------------------------------------------
	max: 47.131MB

📂 Allocator type distribution:
	 MALLOC: 4825088
	 POSIX_MEMALIGN: 1029064
	 CALLOC: 715375
	 REALLOC: 116697
	 MMAP: 183

🥇 Top 5 largest allocating locations (by size):
	- mask_to_rle_pytorch:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:84 -> 662.692MB
	- mask_to_rle_pytorch:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:91 -> 662.630MB
	- load_tensor:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/serialization.py:1357 -> 358.687MB
	- __init__:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/modules/linear.py:96 -> 339.159MB
	- <stack trace unavailable> -> 180.832MB

🥇 Top 5 largest allocating locations (by number of allocations):
	- <stack trace unavailable> -> 936383
	- forward:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/modules/linear.py:114 -> 300182
	- cat:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/segment_anything/utils/amg.py:64 -> 256355
	- batched_mask_to_box:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:51 -> 256253
	- batched_mask_to_box:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:43 -> 250693
```

</p>
</details> 

<details><summary>Details: memray stats memray-output-prefilter-amg-vit_l.bin</summary>
<p>

```
📏 Total allocations:
	7132819

📦 Total memory allocated:
	6.366GB

📊 Histogram of allocation size:
	min: 0.000B
	----------------------------------------------
	< 5.000B   :   78073 ▇
	< 35.000B  : 3339967 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 210.000B : 2305414 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 1.221KB  : 1132402 ▇▇▇▇▇▇▇▇▇
	< 7.260KB  :  226257 ▇▇
	< 43.175KB :   44831 ▇
	< 256.735KB:    3280 ▇
	< 1.491MB  :    2362 ▇
	< 8.865MB  :      68 ▇
	<=52.717MB :     165 ▇
	----------------------------------------------
	max: 52.717MB

📂 Allocator type distribution:
	 MALLOC: 5104480
	 POSIX_MEMALIGN: 1094798
	 CALLOC: 800617
	 REALLOC: 132744
	 MMAP: 180

🥇 Top 5 largest allocating locations (by size):
	- load_tensor:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/serialization.py:1357 -> 1.164GB
	- __init__:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/modules/linear.py:96 -> 1.140GB
	- mask_to_rle_pytorch:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:84 -> 710.521MB
	- mask_to_rle_pytorch:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:91 -> 710.460MB
	- _deepcopy_list:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/copy.py:206 -> 355.178MB

🥇 Top 5 largest allocating locations (by number of allocations):
	- <stack trace unavailable> -> 969540
	- forward:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/modules/linear.py:114 -> 308593
	- batched_mask_to_box:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:51 -> 297101
	- batched_mask_to_box:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:43 -> 284254
	- filter:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/segment_anything/utils/amg.py:49 -> 263420
```

</p>
</details> 

<details><summary>Details: memray stats memray-output-prefilter-amg-vit_h.bin</summary>
<p>

```
📏 Total allocations:
	7838182

📦 Total memory allocated:
	8.850GB

📊 Histogram of allocation size:
	min: 0.000B
	----------------------------------------------
	< 5.000B   :   81028 ▇
	< 35.000B  : 3697177 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 211.000B : 2532510 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 1.233KB  : 1241268 ▇▇▇▇▇▇▇▇▇
	< 7.355KB  :  240018 ▇▇
	< 43.857KB :   40360 ▇
	< 261.478KB:    3139 ▇
	< 1.522MB  :    2385 ▇
	< 9.076MB  :      84 ▇
	<=54.114MB :     213 ▇
	----------------------------------------------
	max: 54.114MB

📂 Allocator type distribution:
	 MALLOC: 5610353
	 POSIX_MEMALIGN: 1236785
	 CALLOC: 862862
	 REALLOC: 128001
	 MMAP: 181

🥇 Top 5 largest allocating locations (by size):
	- load_tensor:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/serialization.py:1357 -> 2.388GB
	- __init__:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/modules/linear.py:96 -> 2.359GB
	- mask_to_rle_pytorch:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:84 -> 723.089MB
	- mask_to_rle_pytorch:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:91 -> 723.028MB
	- _deepcopy_list:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/copy.py:206 -> 253.180MB

🥇 Top 5 largest allocating locations (by number of allocations):
	- <stack trace unavailable> -> 1204709
	- batched_mask_to_box:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:51 -> 332085
	- batched_mask_to_box:/Users/genevieb/Documents/GitHub/micro-sam/micro_sam/_vendored.py:43 -> 320806
	- filter:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/segment_anything/utils/amg.py:49 -> 318677
	- forward:/Users/genevieb/mambaforge/envs/test-micro-sam-mps-nightly/lib/python3.10/site-packages/torch/nn/modules/linear.py:114 -> 318255
```

</p>
</details> 

</p>
</details>

**Comment by constantinpape** [comment]
Hi @GenevieveBuckley,
I think this is resolved now after merging #190. And I think we have a consistent speed up with MPS now!
(If there is anything else here from your side, then feel free to reopen!)
